### ISSUE TITLE
changing mime setting requires forced recompilation of a dep

### ISSUE THREAD

**Issue by sasa1977** [elixir-plug/mime#27]
Mime currently [reads user's setting during compilation](https://github.com/elixir-plug/mime/blob/3cc82cf0918de21c710dd8ba748f038c17e31bd3/lib/mime.ex#L47).

While this makes the interface convenient, it has a serious downside that when user changes their setting, the change won't take effect until a dep is manually recompiled. To make matters worse, the change is silently ignored, and so it might silently make it to production.

Admittedly, docs mention this issue, but force dep recompile isn't really simple in any automated workflow. For example, at my workplace we have a couple of different servers doing automated builds, and to speed up the work, they all keep the cached version of the last build, so changing the configuration won't make it neither to tests, nor to production. The only solutions for this I can think of:

1. Add some trickery in the user project to always force build mime dep
2. Add some more elaborate trickery to detect that mime configuration has been changed since the last time, and then force build mime dep
3. Do manual refresh along the lines of "Nobody should merge until I manually clean cache on every server, push the change, and make sure that the change has been properly propagated".

None of this seems like a compelling option.

As an added downside, the current design makes it impossible to have separate configurations for separate parts of my system (e.g. the main site and an admin site). Both Plug and Phoenix implicitly rely on the global setting cached during the last build of the mime dep.

I'd like to see a more flexible solution here which at the very least properly propagates user's changes to the runtime without requiring manual recompilation. It would also be nice if globalness was dropped.

One way I can think of is to move generation to the user's code. Here's a quick prototype which relies on the existing code to support `extensions/1` and `type/1`:

```elixir
iex> defmodule ConfigurableMime do
  defmacro __using__(user_type_to_exts) do
    quote bind_quoted: [user_type_to_exts: user_type_to_exts] do
      for {type, exts} <- user_type_to_exts do
        def extensions(unquote(type)), do: unquote(exts)
      end

      def extensions(type), do: MIME.extensions(type)

      for {type, exts} <- user_type_to_exts,
          ext <- exts do
        def type(unquote(ext)), do: unquote(type)
      end

      def type(file_extension), do: MIME.type(file_extension)
    end
  end
end

iex> defmodule MyMime do
  use ConfigurableMime, %{"application/foo" => ["bar"]}
end

iex> MyMime.extensions("application/gzip")
["gz", "tgz"]

iex> MyMime.extensions("application/foo")
["bar"]
```

I'm not saying that this is necessarily the best way, but at least it solves both mentioned issues, while still keeping the compiled time guarantees, and without generating multiple copies of the default map in the user's code. If Plug was adapted to explicitly accept the callback module for custom mime configuration, it would be even straightforward to supply parameters at runtime by fetching them from external sources.

**Comment by josevalim** [comment]
Generally speaking, if the solution is to push complexity to the library
users, then it will be unlikely for me to personally accept it.

Regarding globalness, today at least the final user is responsible for
deciding what formats are accepted. We need to be careful to not segment
into multiple mime type modules, causing code/memory bloat, and splitting
your application into independent and hard to configure mime modules.

Anyway, I acknowledge there is an issue, especially regarding compile-time.
For the global aspect, the solution requires careful design. A mixture of
global plus configurable local may work better.
-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by sasa1977** [comment]
> Generally speaking, if the solution is to push complexity to the library users, then it will be unlikely for me to personally accept it.

In principle I agree with that, however I think the issue is more nuanced. One person's complexity push might be another's person freedom of choice.

So for example, putting aside the compile-time issue, the current global design of mime, makes it very difficult to make local variations. If I want to do that, I need to fork not only mime, but also plug, and make some non-trivial changes. That's going to be a nightmare to maintain.

> A mixture of global plus configurable local may work better.

While this support both worlds, my fear is that it strongly favours one over the other, and thus makes it more likely that intermediate libraries (such as e.g. plug) will just support the global setting, which leads us to the same issue described above. If I want to have local variations, I might not need to fork mime, but I might need to fork plug.

So I'd argue that sometimes aiming for client's convenience in the common case, also might make a less common case very difficult to support. Perhaps that's a good choice in this particular situation, but at least I think the trade-off should be acknowledged.

Personally, I think that mime is a low-level specialized library which shouldn't concern itself with user convenience at all, but should instead aim to be flexible so it can support a wide range of scenarios. In other words, I don't think that support for global configuration belongs to mime at all.

I do however agree that life should be made simple for the majority of users. My assumption is that most users use mime exclusively via  Phoenix/Plug, and some of them want to provide their own mime mapping. Therefore it makes sense to make that scenario easy, maybe by providing a plug:

```
plug Plug.MIME, %{"application/foo" => ["foo"], ...}
```

I'd argue that this is in fact simpler and more convenient than moving mime definition to config scripts. Now, this approach has one downside, which leads me to your next point.

> Regarding globalness, today at least the final user is responsible for deciding what formats are accepted. 

With the proposed interface, this benefit disappears. However, if I'm right in assuming that most users just use mime via Phoenix/Plug, then this is not a frequent scenario. 

In addition, I believe that this infrequent scenario shouldn't cause any confusion. Suppose that for some reason I want to use plug and raxx, and provide my own mime configuration. I think it should be obvious that by configuring plug, I didn't solve the issue in raxx.

> causing code/memory bloat

The main reason why I included that particular proposal is because it doesn't cause either of those. The common settings are compiled once and baked in the mime binaries, while only variations sit in the user's code.

However, I personally prefer a different interface which doesn't require callback module at all. Something along the lines of:

```elixir
defmodule MIME do
  @opaque compiled :: # ...

  @spec compile(user_spec :: map) :: compiled
  @spec extensions(String.t, user_spec :: compiled) :: [String.t]
  @spec type(String.t, user_spec :: compiled) :: String.t
end
```

This pushes more complexity to the user, but I believe that in most cases a direct user of mime are other libraries, such as plug and phoenix. Therefore, I'd also say that this interface leaves a lot of freedom to libraries, which can then choose to make the lives of their clients simpler.

Finally, if you say that supporting global configuration in a library such as this one is a good choice, then I have to wonder should every library support global configuration to avoid pushing complexity to users? After all, my impression is that mime is a quite simple and narrow focused library. So if such library promotes global config, isn't it then justified for more complex libraries to do that too? Doesn't global setting approach become a good and recommended pattern?

Personally, I don't like that direction, because we'll end up with a bunch of global/singleton friendly libraries with little to no flexibility at all. 

So in summary (and I apologize for the long rant), I think that global configuration shouldn't be a concern of mime at all, and that end user convenience should be provided by higher-level libraries such as plug and phoenix.

**Comment by josevalim** [comment]
> So for example, putting aside the compile-time issue, the current global design of mime, makes it very difficult to make local variations.

Can you give me one example of when you would want to do that?

This library is modelling a **global database of immutable, read-only information**. All mime-types that you are expecting to receive in your application is known upfront. This library is fully optimized for this case. The current design allows us to extend this global database regardless of the users of the library, leaving the final choice up to the application.

> This pushes more complexity to the user, but I believe that in most cases a direct user of mime are other libraries, such as plug and phoenix. 

Then it is up to Plug and Phoenix to define this contract, not up to this library. That's like saying the `Math` module doesn't expose the interface for someone building a `Calculator`.

> In addition, I believe that this infrequent scenario shouldn't cause any confusion. 

Configuring mime is frequent because many APIs expect versioned or semantic content-types. So for example, you may expect "application/json+v1-app". But again, you know the data types you expect, so you can add those before hand.

> Finally, if you say that supporting global configuration in a library such as this one is a good choice, then I have to wonder should every library support global configuration to avoid pushing complexity to users? 

C'mon. You know very well I am not saying this. Do we really need the straw man?

In case it is not clear: my point is that for the kind of data we are modelling here, compile time configuration is perfectly fine. But if people want to go ahead and remove all of the context from this discussion to push global configuration, there is nothing I can do.

I don't agree with the witch hunt on global configuration. There are valid use cases and this is one of them. Another scenario where I would employ the same solution is something like a database of status codes, where you know upfront which statuses you will be sending back to clients. Having the list of status codes hardcoded with a global configuration to augment them in the few cases necessary is a perfectly fine solution.

**Comment by michalmuskala** [comment]
> So for example, putting aside the compile-time issue, the current global design of mime, makes it very difficult to make local variations. If I want to do that, I need to fork not only mime, but also plug, and make some non-trivial changes. That's going to be a nightmare to maintain.

Are there any concrete examples of when you'd want divergent mime declarations in an application or is that a theoretical concern?

**Comment by sasa1977** [comment]
> Can you give me one example of when you would want to do that?

Suppose I have a special own format which I'm using only in the admin part of my system, and I don't want to support that in the publicly open part.

> This library is modelling a global database of immutable, read-only information. All mime-types that you are expecting to receive in your application is known upfront. This library is fully optimized for this case. The current design allows us to extend this global database regardless of the users of the library, leaving the final choice up to the application.

I'm actually fine with this, but I think that these assumptions should be explained explicitly in the documentation, for the reason I'll state a bit later.

> Configuring mime is frequent because many APIs expect versioned or semantic content-types.

The infrequency I referred to was the situation where at the same time I use two libraries which are using MIME (so e.g. plug and raxx). I'm not arguing that configuring mime is not frequent.

> C'mon. You know very well I am not saying this. Do we really need the straw man?

That's literally the impression I got. However, when you explained the constraints, then it makes a huge difference. I believe it's important to mention these constraints/assumptions in docs, because it communicates the reasons for particular implementation, and the associated trade-offs involved.

I think that many people turn to the "official" libraries (i.e. the ones made by the core team) for inspiration. If they see that mime is using e.g. boot-time app config, they might just copy that approach to their own project. So I think it's important to explain the reasons which led to such implementation.

> I don't agree with the witch hunt on global configuration. There are valid use cases and this is one of them. 

Personally, I think that Logger is a valid case, while I'm not sold on this. As I said, I think the vast majority of mime usages will involve a single Phoenix endpoint, and so you can get the same (if not better) convenience by configuring with a plug. Not only will such interface be simpler, it will also eliminate the boot time configuration, and hence be more runtime friendly.

In other words, by tackling end-user convenience in plug, you can eliminate the global state completely, have a more convenient interface for the most common case (using mime through a single endpoint), and support a wider range of scenarios, including run-time configuration.

> Another scenario where I would employ the same solution is something like a database of status codes

I have the same thoughts regarding this example.

We can agree to disagree here, and I'm fine with that as long as the assumptions you mentioned are explicitly documented.

**Comment by josevalim** [comment]
> Suppose I have a special own format which I'm using only in the admin part of my system, and I don't want to support that in the publicly open part.

The fact you know a mime does not mean that you accept it nor that you have to accept it throughout your application. This lib ships with the CSV mime and many applications do not handle CSV at all. Or they have only certain endpoints that handle CSV, maybe even only behind their admin panel.

In other words, there is absolutely no issue in adding your admin mimes to the list.

> I believe it's important to mention these constraints/assumptions in docs, because it communicates the reasons for particular implementation, and the associated trade-offs involved.

I will be glad to improve this documentation but I want to make it very clear it is not our responsibility if folks copy the design decisions of libraries without asking or trying to understand the context that led to those decisions.

Improving the documentation helps folk understand but those questions should be asked regardless if documentation exists or not.

**Comment by sasa1977** [comment]
> The fact you know a mime does not mean that you accept it nor that you have to accept it throughout your application. 

Sure, but by needing to store this into a global k-v space, there's some chance of an accidental collision (i.e. I choose the name or an ext which is already chosen). With scoped mapping, I can have a better control, perhaps even whitelist only those particular types I'm interested in, to reduce the chance of a mistaken type.

I'm curious though, which scenarios do you feel global config supports which are not convenient to express with something like `plug Plug.Mime, user_settings`?

**Comment by josevalim** [comment]
> Sure, but by needing to store this into a global k-v space, there's some chance of an accidental collision (i.e. I choose the name or an ext which is already chosen).

The likelihood of this happening is truly minimal and, because today you can only configure extensions in a single place, i.e. in your config file, any conflict will be clearly apparent.

> I'm curious though, which scenarios do you feel global config supports which are not convenient to express with something like plug Plug.Mime, user_settings?

The problem with local config is that it is not extensible unless you surface it all upstream. Imagine for example I am working on a controller that handles file uploads. In there I call `plug Plug.Mime, %{"my" => "settings"}`. I put this package on Hex.

Now other users want to use my package and they do "application/hjson+v1+app", which is JSON annotated to their domain. They need to teach my controller to handle it. Because the configuration is local, now the package author needs to surface it upstream. And in this case in particular, since plugs are compile time, we are back to the same problem.

This is why the discussion is much more nuanced than global vs local. The way it is designed today, it is at least global with a unique point for resolution. I would prefer that to having 5 different local points, which may or may not be configurable, and end-up configuring the same extensions all over different places. Given the nature of the data, I still think global w/ unique point for resolution is a very reasonable trade-off to make.

**Comment by josevalim** [comment]
> Perhaps I need to serve some client code over which I don't have the control, so I need a different mapping between a type and an ext.

We are talking about hypothetical scenarios that I haven't heard yet in practice. I don't expect to hear about everything that happens in the community but, as the maintainer of Mime, Plug and Phoenix, I would expect to hear about those cases when users report bugs.

And even if it those cases happened, it is likely you could still tell Plug and Phoenix how to behave. I know Phoenix has explicit APIs to override the inferred (or to be inferred) format from the content type.

Give me an actual case, from someone actually running into this limitation, and I will gladly discuss. We can't optimize for unknowns, this lib is optimized for knowns.

> This is something you can't even do with global config. A dep can't even provide default mime entries - it has to ask the client to manually add them into the global config

Agreed and to me it is a benefit. I have full control over what my app accepts and exposes. No small decisions being scattered around out of my control.

> For a library, I think the discussion is nowhere near as nuanced. 

Alright. So I guess all this discussion here has been for nothing. :) If us spending a whole morning discussing this topic is not nuance, I don't know what is.

Being a library does not mean you should never access the application environment. It means that those cases have to be warranted.

> Those who end up with 5 local points will be the ones who in fact need it, and then locality will actually be a good thing helping them get where they need to be.

If all we have is a local config, then people will abuse the local config, forcing library users to make local choices when a unique global choice would have been completely enough.

Look I agree that local is generally better. I am fine if you value the pros/cons of local more than the pros/cons of global for **this particular case**. But I don't see this moving forward if you can't see that the local approach also has cons. Take protocols for example, it is about a global approach that is extensible, designed explicitly to solve a limitation of local pattern matching. Putting the `local` label on something does not mean it is immediately better than everything else without considering scenarios and use cases.

**Comment by sasa1977** [comment]
> We are talking about hypothetical scenarios

Perhaps it is, but at least it is a concrete scenario which can't be solved with global and can be solved with local. I haven't seen the opposite example in this discussion.

> I have full control over what my app accepts and exposes. No small decisions being scattered around, out of my control.

This is a double-edged sword though. In a sufficiently large app, I usually want to have some decisions scattered, because I can't fit everything into my head. 

Note though that I'm still saying that most usages will have exactly one place where mapping is defined, and that's going to be the endpoint. But the locality will allow them to make different decisions when they want to. 

In contrast, globality simply doesn't cover such cases. It forces absolutely everyone to always have the global configuration. That's perhaps fine, but I'd like to see the concrete reasons for such constraint.

> If all we have is a local config, then people will abuse the local config, forcing library users to make local choices when a unique global choice would have been completely enough in this particular case.

This sounds quite vague. Can you provide a concrete example?

Most people (me included) just start with the usual approach, which is e.g. proposed by generators, or the main info in the doc, and take it from there. Most of us would still end up with a single place where mapping would be defined. And that's a big gain of locality. You can always implement global on top of local, but the other way around won't work.

So by choosing global, you say that no one will ever need to make local variations. That's a big claim, and I'd like to see it supported by some concrete scenario where global brings significant benefits over local. In this particular case, of course.

> But I don't see this moving forward if you can't see that the local approach also has cons in this particular case.

You could consider a Plug.Mime to be global in the scope of an endpoint. So I wasn't really suggesting that users should inject a bunch of mime invocations all over the place in their codebase.

At the end of the day, this library is basically a string-to-string KV. My question is why does it have to be global? So far I fail to see any significant benefit which global brings us for this library. Why make that decision in mime, when you could generate `plug Plug.Mime` with e.g. `phx.new` and have the similar global feel, but with the flexibility of a local configurability for the admittedly rare, but still possible cases where people need it?

**Comment by josevalim** [comment]
> At the end of the day, this library is basically a string-to-string KV. My question is why does it have to be global? So far I fail to see any significant benefit which global brings us for this library. 

Because that's the whole point of the library. To provide a **global and performant lookup of the mime database**.

Having a global library does not block libraries from providing local choices. And, as a matter of fact, **they do**:

  * In Phoenix: https://github.com/phoenixframework/phoenix/blob/3847705cd7161b972f4d5856b36ccd5e62f86f0d/lib/phoenix/controller.ex#L817-L829
  * In Plug: https://github.com/elixir-plug/plug/blob/10948b83a4441d4ef0f09a7563d307b1d3903080/lib/plug/static.ex#L87-L89

But we have 1855 known content-types, how can we make them available by default for users? We embed them into a database that is compiled into memory for performance.

Therefore, the goal of this library is to provide a copy of the MIME database for applications that want to provide convenient defaults for mime types. Those libraries are free (and recommended) to provide local choices too.

As I said, I don't agree on the witch hunt on global configuration. Having a global configuration here means that I can extend all uses of Plug.Static in my system transparently. The ability to use local choices only when relevant allows us to mix and match those approaches when desired.

Make of that what you will. I have no plans to change how this library works and I have already poured enough time and energy into this discussion. The decision process has already been outlined, others can agree or disagree.

**Comment by sasa1977** [comment]
> I don't agree on the witch hunt on global configuration

For me, this is not a witch hunt on global config. I'm not saying global config is universally bad, but it has its constraints, and if powered by app env, it can have more limitations. That's why I was questioning it here, as I would do with any other library if I opened up a related issue. I'm personally not convinced that its usage is justified in this case, but obviously we'll have to agree to disagree.
